### ISSUE TITLE
Make the syntax of meta consistent with attributes, links, etc

### ISSUE THREAD

**Issue by Erol** [Netflix/fast_jsonapi#292]
Credit to @manojmj92 for adding this feature.

Just wanted to explore the possibility of making the syntax of `meta` consistent. Currently, we pass a name whenever we define an `attribute` or `link`, like:

```ruby
class PersonSerializer
  attribute :name
  link :self
end
```

And `meta` takes in a block:

```ruby
class PersonSerializer
  meta do
    { some: value }
  end
end
```

Would it be worth pursuing making the `meta` syntax to:

```ruby
class PersonSerializer
  meta :some_meta_field
  meta :another_meta_field do |object|
    ...
  end
end
```

What are your thoughts about this?

**Comment by manojmj92** [comment]
@Erol thanks for raising this!

While I like the idea, the case of `meta` is a little different from the case of `attribute` or `link` you have mentioned.

Eg: When you say 

```ruby
attribute :name
attribute :address
```

it actually creates 2 different keys with the name `name` and `address`.

While in the case of `meta`, no matter how many `meta` keys you specify like 

```
class PersonSerializer
  meta :some_meta_field
  meta :another_meta_field do |object|
    ...
  end
end
```
there can only be one and only one meta key in the response.

And since there can only be one meta key, how do we collect all the meta data you have mentioned?

If `some_meta_field` is an array and `another_meta_field` is a hash, how do you contact them? Do you put the array inside the hash or the hash inside the array? It leads to unnecessary confusion.

So I strongly feel having the ability to add multiple `meta` declarations gives no value add to us.

Your thoughts?

**Comment by Erol** [comment]
You're right that there can only be one `meta` key per record. But that is also the case when using `attribute` in the sense that there can only be one `attributes` block per record.

I was thinking that all keys defined by the `meta` directive will be placed directly inside the `meta` object, similar to how `attribute` does it for the `attributes` object.

Here's an example with both `attribute` and `meta`:

```ruby
class BookSerializer
  attribute :name
  attribute :description
  meta :copyright # Will use Book#copyright, similar to how attribute currently behaves
  meta :authors do
    %w[Arnold Bob]
  end
end

BookSerializer.new([person]).serialized_json
```

This will create the following JSON:

```json
{
  "data": [
    {
      "id": "1",
      "type": "book",
      "attributes": {
        "name": "...",
        "description": "..."
      },
      "meta": {
        "copyright": "...",
        "authors": [
          "Arnold",
          "Bob"
        ]
      }
    }
  ]
}
```

This is also consistent with the `link` directive.

**Comment by manojmj92** [comment]
Hey @Erol, thanks for the detailed explanation, I get your point now.
Looks like a good addition, and I can implement this.

But I'll wait on @shishirmk's opinion before moving forward.

**Comment by manojmj92** [comment]
Also @Erol, one good thing about your idea is that we get to override the methods used for `meta` tag when serializers are inherited. Using the block style hinders us from doing so. :)

**Comment by Erol** [comment]
No worries, looking back I should have given a detailed example. I wrote the initial post in haste 🙂 

Yeah that's an additional benefit, we can inherit and choose to override per key as needed.

**Comment by Erol** [comment]
Any feedback on this @shishirmk?

**Comment by Genkilabs** [comment]
Actually, what I would really like for my use case is per-item meta in a collection. Per the JSONAPI spec: 
```a resource object MAY contain any of these top-level members:```
```
- attributes: an attributes object representing some of the resource’s data.
- relationships: a relationships object describing relationships between the resource and other JSON API resources.
- links: a links object containing links related to the resource.
- meta: a meta object containing non-standard meta-information about a resource that can not be represented as an attribute or relationship.
```
SerializationCore already includes a pattern for `relationships_hash`, `attributes_hash`, and `links_hash`. So I would propose a similar pattern for the `meta_hash` on an individual record.

I envision adding per-object meta working similarly to links https://github.com/Netflix/fast_jsonapi#links-per-object

...I know I kind of scope creeped there, but if there are changes happening to meta, its worth considering.

**Comment by manojmj92** [comment]
@Genkilabs this is something that I have already implemented and merged in the `dev` branch. It will be live in version 1.4 of this gem. PR: #280

**Comment by shishirmk** [comment]
Released in [Version 1.4](https://github.com/Netflix/fast_jsonapi/releases/tag/1.4)
